src/Pipeline.cpp

```diff
@@ -70,6 +70,23 @@ std::string sanitize_function_name(const std::string &s) {
     return name;
 }
 
+class FindVariablesWithUnsetParams : public IRVisitor {
+    using IRVisitor::visit;
+
+    void visit(const Variable *op) override {
+        if (op->param.defined() && !op->param.is_buffer() && op->param.name() != "__user_context") {
+            user_assert(op->param.has_scalar_value())
```

Comment:
I don't think this error message is right. Won't it trigger on all attempts to realize a pipeline with unset params, regardless of whether or not you're in a generator? I think the much more likely scenario is someone using the JIT forgetting to set the value of a Param<T>. Also, you can absolutely compile_to_callable without setting the scalar params.

Reply:
> Won't it trigger on all attempts to realize a pipeline with unset params, regardless of whether or not you're in a generator?

Yes. That's a feature, not a bug.

> I don't think this error message is right.

Yeah, this is a bit confusing -- IIRC the reason for wording it this way is that the "other" cases (realize, callable) end up being caught via `read_only_scalar_address()` noting the invalid value and failing (apparently they don't add the interposed Variable, I don't recall why, it's been a few days).